Adds section on Verified commits (#152)
MetOffice/git-novice: learners/setup.md

:::

## Signed/Verified Commits

Using GPG, or SSH, you can [sign tags and commits](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification) locally.
These tags or commits are marked as verified on GitHub

> so other people can be confident that the changes come from a trusted source.

Reviewer: ```suggestion
so other people can be confident that the changes come from you.
```
The signing allows you to prove you made a commit. It is up the the receiver whether or not they trust you. A malicious third-party can sign their commits, but that doesn't mean those commits should be trusted.

Author: I did wonder if this was too vague (It's straight from the GitHub docs!) I agree lets change it.

Some repositories require all commits to be
signed and verified before they can be merged into the main branch.

Choose whichever method you prefer for signing commits, GPG or SSH.

> GPG is more widely used for signing commits,

Reviewer: ```suggestion
GPG is more traditionally used for signing commits,
```
GPG might well be less common nowadays. It was historically the only option, but now all systems should support SSH signing.

That said, GPG does have some extra features, like expiration and trust levels, that make it more suited to distributed development. Using a central forge like GitHub obviates the need for most of these features.

but SSH is easier to set up and maintain.
GPG keys can be set to expire.
SSH keys do not expire but can be revoked by removing the public key from GitHub.

::: tab

### SSH

To sign commits with SSH follow the instructions
in the [SSH commit signature verification documentation](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification#ssh-commit-signature-verification).

Although you can use the same SSH key for both authentication and signing,

> it is recommended to use a separate key for signing commits.

Author: ```suggestion
You can use the same SSH key for both authentication and signing,
or create a new SSH key for signing commits.
```

Author: @jfrost-mo Is this ok? Would resolve our other conversation.

Reviewer: Yep, that would sort it. I don't like putting recommendations that can't be supported by reasoning.

> it is recommended to use a separate key for signing commits.

Reviewer: Should we provide a bit of rational for why we recommend to use separate keys?

The main reason I can think of is that it allows you to use a passphrase on your signing key without needing one on your authentication key. This is useful if you need to be able to clone or pull non-interactively.

Author: @jfrost-mo Yes also you could copy just the signing key across platforms if it's separate.
My main point would be one key for one use, if you want to remove your ssh auth key from GitHub and then also delete it locally without remembering you're also using it for signing it could cause a headache.

Perhaps we should recommend a name for the key `~/.ssh/id_ed25519_github_signing` like we do the auth key.

Reviewer: Perhaps, though I'm a little tempted to drop the recommendation entirely for simplicity in these docs. GitHub now verifies signatures at push time, so even if you accidently delete your key it only means you need to upload a new one. The only commits that would be affected are those you have committed but not pushed, which should be fairly rare.

### GPG

To sign commits with GPG follow the instructions
in the [GPG commit signature verification documentation](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification#gpg-commit-signature-verification).

:::

> Optionally, to configure Git to sign all commits and tags by default,

Reviewer: ```suggestion
To configure Git to sign all commits and tags by default,
```

I'd highly suggest we phrase this as required, as signatures won't harm projects that don't need them, while forgetting to do the signing is a bit annoying to fix (requiring rebasing all your commits).
